plugins/woocommerce/includes/wc-product-functions.php

```diff
@@ -158,6 +173,67 @@ function wc_delete_product_transients( $post_id = 0 ) {
 	do_action( 'woocommerce_delete_product_transients', $post_id );
 }
 
+/**
+ * Asynchronously delete related product transients when a product is changed.
+ * This is necessary because changing one product might affect many related products.
+ *
+ * @since 9.8.1
+ * @param array $post_id Arguments passed from the async scheduler.
+ */
+function wc_delete_related_product_transients( $post_id ) {
+	global $wpdb;
+
+	if ( ! is_numeric( $post_id ) ) {
+		return;
+	}
+
+	$transient_name       = 'wc_related_' . $post_id;
+	$old_transient        = get_transient( $transient_name );
+	$old_related_products = array();
+
+	if ( is_array( $old_transient ) ) {
+		$old_related_products = $old_transient[ array_key_first( $old_transient ) ];
+	}
+
+	// Delete current product transient.
+	delete_transient( $transient_name );
+
+	// Gets new related products and sets current product transient.
+	$new_related_products = wc_get_related_products( $post_id, 1000 );
+
+	// Combine all product IDs that need their transients cleared.
+	$related_product_ids = array_unique(
+		array_merge(
+			$old_related_products,
+			$new_related_products
+		)
+	);
+
+	if ( empty( $related_product_ids ) ) {
+		return;
+	}
+
+	// Create the list of transient names to delete.
+	$transient_names = array();
+	foreach ( $related_product_ids as $id ) {
+		$transient_names[] = '_transient_timeout_wc_related_' . $id;
+		$transient_names[] = '_transient_wc_related_' . $id;
+	}
+
+	// Create placeholders for the IN clause.
+	$placeholders = implode( ',', array_fill( 0, count( $transient_names ), '%s' ) );
+
+	// Delete specific timeout and transient entries in a single query.
+	$wpdb->query(
+		$wpdb->prepare(
+			"DELETE FROM {$wpdb->options} WHERE option_name IN ($placeholders)",
+			$transient_names
+		)
+	);
```

Comment:
Not all transients are stored in the database. We can implement a check for this like the below (haven't verified it though)

```suggestion
	// Delete specific timeout and transient entries.
	if ( wp_using_ext_object_cache() ) {
		// When using external object cache, we need to delete each transient individually
		foreach ( $related_product_ids as $id ) {
			delete_transient( 'wc_related_' . $id );
		}
	} else {
		// For database storage, we can use a single query for better performance
		$wpdb->query(
			$wpdb->prepare(
				"DELETE FROM {$wpdb->options} WHERE option_name IN ($placeholders)",
				$transient_names
			)
		);
	}
```